net/captivedetection/endpoints.go

```diff
@@ -0,0 +1,213 @@
+	// because they do not depend on DNS resolution.
+	for _, region := range derpMap.Regions {
+		if region.Avoid {
+			continue
+		}
+		for _, node := range region.Nodes {
+			if node.IPv4 == "" {
+				continue
+			}
+			str := "http://" + node.IPv4 + "/generate_204"
+			u, err := url.Parse(str)
+			if err != nil {
+				logf("captivedetection: failed to parse DERP node URL %q: %v", str, err)
+				continue
+			}
+			p := DERPMapOther
+			if region.RegionID == preferredDERPRegionID {
+				p = DERPMapPreferred
+			}
+			e := Endpoint{u, http.StatusNoContent, "", true, p}
+			endpoints = append(endpoints, e)
+		}
+	}
+
+	// Let's also try the default Tailscale coordination server and admin console.
+	// These are likely to be blocked on some networks.
+	appendTailscaleEndpoint := func(urlString string) {
+		u, err := url.Parse(urlString)
+		if err != nil {
+			logf("captivedetection: failed to parse Tailscale URL %q: %v", urlString, err)
+			return
+		}
+		endpoints = append(endpoints, Endpoint{u, http.StatusNoContent, "", false, Tailscale})
+	}
+	appendTailscaleEndpoint("http://controlplane.tailscale.com/generate_204")
+	appendTailscaleEndpoint("http://login.tailscale.com/generate_204")
+
+	// Lastly, to be safe, let's also include some well-known captive portal detection URLs that are not under the
+	// tailscale.com umbrella. These are less likely to be blocked on public networks since blocking them
+	// would break captive portal detection for many devices.
+	appendPlatformEndpoint := func(urlString string, statusCode int, expectedContent string) {
+		u, err := url.Parse(urlString)
+		if err != nil {
+			logf("captivedetection: failed to parse Platform URL %q: %v", urlString, err)
+			return
+		}
+		endpoints = append(endpoints, Endpoint{u, statusCode, expectedContent, false, Platform})
+	}
+
+	switch goos {
+	case "windows":
+		appendPlatformEndpoint("http://www.msftconnecttest.com/connecttest.txt", http.StatusOK, "Microsoft Connect Test")
+		appendPlatformEndpoint("http://www.msftncsi.com/ncsi.txt", http.StatusOK, "Microsoft NCSI")
+	case "darwin", "ios":
+		appendPlatformEndpoint("http://captive.apple.com/hotspot-detect.html", http.StatusOK, "Success")
+		appendPlatformEndpoint("http://www.thinkdifferent.us/", http.StatusOK, "Success")
+		appendPlatformEndpoint("http://www.airport.us/", http.StatusOK, "Success")
+	case "android":
+		appendPlatformEndpoint("http://connectivitycheck.android.com/generate_204", http.StatusNoContent, "")
+		appendPlatformEndpoint("http://connectivitycheck.gstatic.com/generate_204", http.StatusNoContent, "")
+		appendPlatformEndpoint("http://play.googleapis.com/generate_204", http.StatusNoContent, "")
+		appendPlatformEndpoint("http://clients3.google.com/generate_204", http.StatusNoContent, "")
+	default:
+		appendPlatformEndpoint("http://detectportal.firefox.com/success.txt", http.StatusOK, "success")
+		appendPlatformEndpoint("http://network-test.debian.org/nm", http.StatusOK, "NetworkManager is online")
+	}
```

Comment:
I'd probably remove these for this round. Put them in a doc somewhere for now and we can revisit later.